pkg/controller/usersignup/usersignup_controller_test.go

```diff
@@ -0,0 +1,404 @@
+package usersignup
+
+import (
+	"context"
+	"github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
+	"github.com/codeready-toolchain/host-operator/pkg/apis"
+	"github.com/codeready-toolchain/host-operator/pkg/config"
+	"github.com/codeready-toolchain/toolchain-common/pkg/cluster"
+	"github.com/operator-framework/operator-sdk/pkg/log/zap"
+	"github.com/satori/go.uuid"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/h2non/gock.v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	fakeclientset "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/kubefed/pkg/apis/core/common"
+	"sigs.k8s.io/kubefed/pkg/apis/core/v1beta1"
+	"testing"
+)
+
+const (
+	nameMember = "east"
+)
+
+func TestReadUserApprovalPolicy(t *testing.T) {
+	r, _ := prepareReconcile(t, "test")
+
+	// Create a new ConfigMap
+	cmValues := make(map[string]string)
+	cmValues[config.ToolchainConfigMapUserApprovalPolicy] = config.UserApprovalPolicyAutomatic
+	cm := &v1.ConfigMap{
+		Data: cmValues,
+	}
+	cm.Name = config.ToolchainConfigMapName
+
+	_, err := r.clientset.CoreV1().ConfigMaps(config.GetOperatorNamespace()).Create(cm)
+	require.NoError(t, err)
+
+	policy, err := r.ReadUserApprovalPolicyConfig()
+	require.NoError(t, err)
+	require.Equal(t, config.UserApprovalPolicyAutomatic, policy)
+}
+
+func TestUserSignupWithAutoApproval(t *testing.T) {
+	userSignup := &v1alpha1.UserSignup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "foo",
+			Namespace: config.GetOperatorNamespace(),
+			UID: types.UID(uuid.NewV4().String()),
+		},
+		Spec: v1alpha1.UserSignupSpec{
+			UserID: "foo",
+			Approved: false,
+		},
+	}
+
+	r, req := prepareReconcile(t, userSignup.Spec.UserID, userSignup)
+
+	createMemberCluster(r.client)
+	defer clearMemberClusters(r.client)
```

Comment:
@sbryzak Just a note - technically you don't need to setup/mock the whole KubeFedCluster machinery. You could easily just mock the function `cluster.GetMemberClusters()` in the same way as I do it here https://github.com/codeready-toolchain/host-operator/pull/41/files#diff-93c8b7a9cc0875ddd6d3002d3d0b4253R168-R200
With this, you only mock the function (the function is tested in the common repo enough), so there is no need of adding/cleaning the KubeFedCluster CRD as well as adding secrets or mocking http calls. The test is then much simpler I think

Reply:
I agree it makes the test simpler, however it seems to complicate the controller a little more which now requires an additional function reference to be passed in its constructor.  I think I would prefer to keep the additional complexity in the test rather than in the controller.